File: TweetSearchFSharp/Searcher.fs
module TweetSearchFSharp

open System
open System.IO
open System.Text
open System.Net.Http
open System.Net.Http.Headers
open System.Linq
open Newtonsoft;

let key = "" // TODO enter your consumer key here!
let secret = "" // TODO enter your consumer secret here!

let create_basic_auth(key, secret) = 
    sprintf "%s:%s" key secret
    |> Encoding.UTF8.GetBytes
    |> Convert.ToBase64String

let basic_auth =
    (key, secret) |> create_basic_auth

type credentials = {
    access_token : string
}

let deserialize_credentials(raw) =
    Json.JsonConvert.DeserializeObject<credentials>(raw)
    
let get_bearer_token =
    async {
        let http = new HttpClient()
        http.DefaultRequestHeaders.Add("Authorization", "Basic " + basic_auth)
        let content = new System.Net.Http.StringContent("grant_type=client_credentials", Encoding.UTF8, "application/x-www-form-urlencoded")
        let! response = http.PostAsync("https://api.twitter.com/oauth2/token", content) |> Async.AwaitTask
        response.EnsureSuccessStatusCode |> ignore
        let! s = response.Content.ReadAsStringAsync() |> Async.AwaitTask
        let creds = deserialize_credentials(s)
        return creds.access_token
    }

type tweet = {
    text : string
}

type search_response = {
    statuses : tweet []
}

let deserialize_tweets(raw) =
    let resp = Json.JsonConvert.DeserializeObject<search_response>(raw)
    resp.statuses

let get_tweets(search_term) =
    async {
        let http = new HttpClient()
        let! bearer_token = get_bearer_token
        http.DefaultRequestHeaders.Add("Authorization", "Bearer " + bearer_token)
        let! response = http.GetAsync("https://api.twitter.com/1.1/search/tweets.json?q=" + search_term + "&count=100") |> Async.AwaitTask
        response.EnsureSuccessStatusCode |> ignore
        let! s = response.Content.ReadAsStringAsync() |> Async.AwaitTask
        let tweets = deserialize_tweets(s)
        return tweets
    }

File: TweetSearchCSharp/Searcher.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace TweetSearchCSharp
{
    public class Searcher
    {
        private const string Key = ""; // TODO enter your consumer key here!
        private const string Secret = ""; // TODO enter your consumer secret here!
        private readonly string _bearerToken;
        
        private static string BasicAuth
        {
            get { return Convert.ToBase64String(Encoding.UTF8.GetBytes(Key + ":" + Secret)); }
        }

        [DataContract]
        public class Credentials
        {
            [DataMember(Name = "access_token")]
            public string AccessToken { get; set; }
        }

        public static async Task<string> BearerTokenAsync()
        {
            using (var http = new HttpClient())
            {
                http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Basic", BasicAuth);
                var responseMessage =
                    await
                        http.PostAsync("https://api.twitter.com/oauth2/token",
                            new StringContent("grant_type=client_credentials", Encoding.UTF8,
                                "application/x-www-form-urlencoded"));

                return (await Deserialize<Credentials>(responseMessage)).AccessToken;
            }
        }

        public Searcher(string bearerToken)
        {
            _bearerToken = bearerToken;
        }

        [DataContract]
        public class Tweet
        {
            [DataMember(Name = "text")]
            public string Text { get; set; }
        }

        [DataContract]
        public class SearchResponse
        {
            [DataMember(Name = "statuses")]
            public Tweet[] Tweets { get; set; }
        }

        public async Task<Tweet[]> SearchAsync(string term, int count = 100)
        {
            using (var http = new HttpClient())
            {
                http.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", _bearerToken);
                var responseMessage =
                    await
                        http.GetAsync(string.Format("https://api.twitter.com/1.1/search/tweets.json?q={0}&count={1}",
                            Uri.EscapeDataString(term), count));

                return (await Deserialize<SearchResponse>(responseMessage)).Tweets;
            }
        }

        private static async Task<T> Deserialize<T>(HttpResponseMessage response)
        {
            if (!response.IsSuccessStatusCode)
                throw new Exception(string.Format("HTTP request failed with {0}", response.StatusCode));

            try
            {
                var stream = await response.Content.ReadAsStreamAsync();
                using (var reader = new JsonTextReader(new StreamReader(stream, Encoding.UTF8)))
                {
                    return new JsonSerializer().Deserialize<T>(reader);
                }
            }
            catch (Exception ex)
            {
                throw new Exception(string.Format("Response for {0} could not be deserialized!", response.RequestMessage.RequestUri), ex);
            }
        }
    }
}
